File: solutions/ruby/resistor-color-trio/3/resistor_color_trio.rb
class ResistorColorTrio
  COLORS = %w[black brown red orange yellow green blue violet grey white]

    UNITS = {
    1_000_000_000 => "gigaohms",
    1_000_000     => "megaohms",
    1_000         => "kiloohms"
    }
  
  def initialize(colors)
    @value = colors
      .map { |c| COLORS.index(c) }
        .then { |a, b, z| (a * 10 + b) * (10 ** z) }
  end

  def resistor
    unit = UNITS.keys.find { |k| @value >= k }
    value = 
      if unit
          "#{@value / unit} #{UNITS[unit]}"
        else
          "#{@value} ohms"
        end
    
  end
  
  def label
    "Resistor value: #{resistor}"
  end
end

File: solutions/c/resistor-color-trio/1/resistor_color_trio.c
#include "resistor_color_trio.h"

#define K 1000
#define M 1000000
#define G 1000000000

resistor_value_t color_code(resistor_band_t colors[]) {
    int value = (colors[0] * 10) + colors[1];

    long long ohms = value;
    for (int i = 0; i < (int)colors[2]; i++) {
        ohms *= 10;
    }

    resistor_value_t result;

    if (ohms >= G) {
        result.value = ohms / G;
        result.unit = GIGAOHMS;
    } else if (ohms >= M) {
        result.value = ohms / M;
        result.unit = MEGAOHMS;
    } else if (ohms >= K) {
        result.value = ohms / K;
        result.unit = KILOOHMS;
    } else {
        result.value = ohms;
        result.unit = OHMS;
    }

    return result;
}